testing/upgradetest/versions.go

```diff
@@ -81,68 +155,84 @@ func getUpgradableVersions(ctx context.Context, vList *tools.VersionList, upgrad
-	mostRecentVersion := sortedParsedVersions[0]
-	mostRecentIsUnreleased := mostRecentVersion.IsSnapshot()
+	return findRequiredVersions(sortedParsedVersions, reqs)
+}
 
-	var upgradableVersions []*version.ParsedSemVer
-	for _, parsedVersion := range sortedParsedVersions {
-		if currentMajorSelected == currentMajorVersions && previousMajorSelected == previousMajorVersions {
-			// we got all the versions we need, break the loop
-			break
-		}
+// findRequiredVersions filters the version list according to the set requirements.
+func findRequiredVersions(sortedParsedVersions []*version.ParsedSemVer, reqs VersionRequirements) ([]string, error) {
+	parsedUpgradeToVersion, err := version.ParseVersion(reqs.UpgradeToVersion)
+	if err != nil {
+		return nil, fmt.Errorf("upgradeToVersion %q is not a valid version string: %w", reqs.UpgradeToVersion, err)
+	}
+	upgradableVersions := make([]string, 0, reqs.CurrentMajors+reqs.PreviousMajors+reqs.PreviousMinors+reqs.RecentSnapshots)
 
-		if !parsedVersion.Less(*parsedUpgradeToVersion) {
-			// skip as testing version is less than version to upgrade from
-			continue
-		}
+	currentMajor := parsedUpgradeToVersion.Major()
+	currentMinor := parsedUpgradeToVersion.Minor()
 
-		isPrevMinor := (parsedUpgradeToVersion.Major() == parsedVersion.Major()) &&
-			(parsedUpgradeToVersion.Minor()-parsedVersion.Minor()) == 1
-
-		if parsedVersion.IsSnapshot() {
-			// Allow returning the snapshot build of the previous minor if the current version is unreleased.
-			// In this situation the previous minor branch may also be unreleased immediately after feature freeze.
-			if !mostRecentIsUnreleased || !isPrevMinor {
-				continue
-			}
-		} else {
-			// Skip the non-snapshot build of the previous minor since it might only be available at
-			// staging.elastic.co which is not a default binary download location.
-			if mostRecentIsUnreleased && isPrevMinor {
-				continue
-			}
-		}
+	currentMajorsToFind := reqs.CurrentMajors
+	previousMajorsToFind := reqs.PreviousMajors
+	previousMinorsToFind := reqs.PreviousMinors
+	recentSnapshotsToFind := reqs.RecentSnapshots
 
-		if parsedVersion.Major() == currentMajor && currentMajorSelected < currentMajorVersions {
-			upgradableVersions = append(upgradableVersions, parsedVersion)
-			currentMajorSelected++
+	for _, version := range sortedParsedVersions {
+		switch {
+		// we skip version above the target
+		case !version.Less(*parsedUpgradeToVersion):
 			continue
-		}
 
-		if parsedVersion.Major() < currentMajor && previousMajorSelected < previousMajorVersions {
-			upgradableVersions = append(upgradableVersions, parsedVersion)
-			previousMajorSelected++
-			continue
+		case recentSnapshotsToFind > 0 && version.IsSnapshot():
+			upgradableVersions = append(upgradableVersions, version.String())
+			recentSnapshotsToFind--
+
+		// for the rest of the checks we capture only released versions
+		case version.Prerelease() != "" || version.BuildMetadata() != "":
+
```

Comment:
If I understood correctly, we don't want pre-release or versions with build metadata, so this `case` entry is here to skip those versions. In that case I believe explicitly adding a `continue` helps to understand the code.
```suggestion
		// for the rest of the checks we capture only released versions
		case version.Prerelease() != "" || version.BuildMetadata() != "":
			continue

```

Reply:
@belimawr I reconsidered and added the `continue` because the switch has a `default` block with `break` anyway. If it reads better with `continue` we should add it, you're right.